File: GA/ga_toolbox/mutationV.m
function mutateP = mutationV(P,lowchance,highchance,LR,UR)
%mutation all position by chance based on cost
% P: ranked population , with the best cost at the fisrt row

mutateP = P;  %[n_pop x n_parameter]
mutateChance = linspace(lowchance,highchance,size(mutateP,1));
mask = rand(size(mutateP)) < repmat(mutateChance',1,size(mutateP,2));
tmp = population(size(mutateP,1),size(mutateP,2),LR,UR);
mutateP(mask)=tmp(mask);


end

File: GA/ga_toolbox/mutationV.py
import numpy as np
from GA.ga_toolbox.population import population


def mutationV(P, lowchance, highchance, LR, UR):
    """
    Mutate all positions by chance, where mutation probability increases
    linearly from best (first row) to worst (last row) of the ranked population.

    Parameters
    ----------
    P          : np.ndarray  [n_pop x n_parameter]  ranked population,
                             best (lowest cost) individual at the first row
    lowchance  : float  mutation probability for the best individual
    highchance : float  mutation probability for the worst individual
    LR         : array-like  lower boundary
    UR         : array-like  upper boundary

    Returns
    -------
    mutateP : np.ndarray  [n_pop x n_parameter]  mutated population
    """
    mutateP = P.copy()                                          # [n_pop x n_parameter]

    mutateChance = np.linspace(lowchance, highchance, mutateP.shape[0])  # per-row probability

    mask = np.random.rand(*mutateP.shape) < mutateChance[:, np.newaxis]  # broadcast

    tmp = population(mutateP.shape[0], mutateP.shape[1], LR, UR)
    mutateP[mask] = tmp[mask]

    return mutateP